doc/ADR/cpp_api_modularisation.md

```diff
@@ -0,0 +1,103 @@
+codebase easier to understand and maintain.
+
+## The Plan Going Forward
+
+Given the above outline, we have reached a point where we are strongly motivated
+to take action to better componentise and modularise CProver.
+
+It's also an opportune time for us, given the existence of `Libcprover-cpp`
+the C++ API that we built to support interfacing with Rust (for now - other
+languages may be coming in the future): we can use this as the basis of
+development for an API exposing the interfaces of the various other modules
+and refactor them into the better-defined shape we want them to take on
+an incremental basis.
+
+Of course, this is a project that is massive in scope, potentially being
+exposed to further scope creep. We acknowledge that any effort to do what
+we have discussed already is going to be a multi-year effort from our end,
+and that we will need community alignment to achieve the outcome we're
+looking for.
+
+This is why we are looking into testing the approach on a smaller component
+first, to get a better feel for the amount of effort and any challenges
+lurking in the dark.
+
+## `goto-bmc` and `libcprover-cpp`
+
+One of the objectives of our modularisation efforts is to decouple the
+various components `CBMC` is based on (front-ends, backends, etc) to allow
+for reuse/recombination. As a first segue into the larger effort, we wanted
+a tool focusing only on running symex (the backend of our analysis engine)
+on a GOTO-binary that has been preprocessed into an appropriate form.
+
+We took the first steps for that in [cbmc#7762](https://github.com/diffblue/cbmc/pull/7762).
+
+The aim of the PR was not only to allow for the tool with the narrower-scope
+to come to life, but also to see if we could expose just enough of the
+process to the C++ API and use that as the basis of the new tool.
+
+This whole process has been very informational: we found out that not only
+we *can* use the C++ API in that capacity, but also that extending the API
+as and when we need to, and doing the various refactorings to the other tools
+on a Just-In-Time basis is viable.
+
+There have been, however, some limitations:
+
+* The C++ API is still nascent, and as such its support for various workflows
+  is just not present (at first). We can (and do) build things fast, as and
+  when we need them - but it's nowhere near feature complete to provide for
+  all of a user's workflows at the time of this writing.
+
+* CProver tools have primarily been based on textual output to report on the
+  results of their function (be it analysis, or transformations, etc). This
+  hasn't been a problem up until this point (with the caveat that occassionally
+  requests for support of new textual formats come up and adding support for
+  those has become a laborious process).
+
+  There's a need however for the separation of concerns between the production
+  of the results by the analysis engine and the presentation layer for those.
+
```

Comment:
I suggest to mention the almost non-existent option support.

Reply:
@NlightNFotis This has not been addressed

Reply:
I'm not sure this is a good idea - my understanding is that this is a temporary limitation. It's not likely to outlive this document, and if we add it now on the basis that we remove this later, we will most likely forget.